### PR TITLE
gitserver: get remote URL from database

### PR DESCRIPTION
**Pull request by keegancsmith**
Currently gitserver stores the remote URL in a repositories
.git/config. It is informed of this URL lazily. Some requests include
the "URL" field. When this field changes, we update the
configuration. This leads to hard to understand semantics in clients,
for example CachedGitRepo, GitRepo and CommandRetryer all exist for
different strategies of obtaining the remote URL.

We now have the CloneURL stored in the DB. So instead of gitserver
sometimes being told the URL, gitserver can directly ask the DB for the
URL. This commit implements that by introducing a connection to the
database in gitserver and updating all call sites which read the URL to
instead ask the database.

Note: Do not merge until production has been updated so that gitserver can speak to cloud-sql-proxy.

Fixes https://github.com/sourcegraph/sourcegraph/issues/14910

Co-authored-by: @tsenart

### CODE REVIEW

File: cmd/gitserver/main.go [hunk 1]
```diff
@@ -50,10 +56,26 @@ func main() {
...
 		DesiredPercentFree:      wantPctFree2,
+		GetRemoteURL: func(ctx context.Context, repo api.RepoName) (string, error) {
+			r, err := repoStore.GetByName(ctx, repo)
+			if err != nil {
+				return "", err
+			}
+			for _, info := range r.Sources {
+				return info.CloneURL, nil
+			}
+			return "", fmt.Errorf("internal error no sources for %s", repo)
```
Comment (ryanslade) [inline cmd/gitserver/main.go:77]: Why do we need to say `internal error`?

```suggestion
			return "", fmt.Errorf("no sources for %q", repo)
```

File: cmd/gitserver/main.go [hunk 2]
```diff
@@ -128,3 +150,36 @@ func parsePercent(s string) (int, error) {
...
 	return p, nil
 }
+
+// getRepoStore initializes a connection to the database and returns a
+// RepoStore.
+func getRepoStore() (*db.RepoStore, error) {
+	//
+	// START FLAILING
+
+	// Gitserver only does internal actor. We rely on the frontend to do authz
```
Comment (ryanslade) [inline cmd/gitserver/main.go:160]:
```suggestion
	// Gitserver is an internal actor. We rely on the frontend to do authz
```

File: cmd/gitserver/main.go [hunk 3]
```diff
@@ -128,3 +150,36 @@ func parsePercent(s string) (int, error) {
 	}
 	return p, nil
 }
+
+// getRepoStore initializes a connection to the database and returns a
+// RepoStore.
+func getRepoStore() (*db.RepoStore, error) {
+	//
+	// START FLAILING
```
Comment (ryanslade) [inline cmd/gitserver/main.go:158]: ?

**Review by ryanslade — APPROVED** [review]

File: cmd/gitserver/main.go (hunk 3 above)
Comment (keegancsmith) [reply cmd/gitserver/main.go:158]: code intel has the same comment around working around authz when accessing the store outside of frontend. Thought I'd preserve it :)

**Review by tsenart — APPROVED** [review]

**Review by unknwon — APPROVED** [review]

**Comment by unknwon** [comment]
Looks a lot nicer.

**Comment by keegancsmith** [comment]
@tsenart I can land and shepherd this into production on Monday morning. Then our next hack session we can work on removing cruft in the client side of this :)

**Comment by tsenart** [comment]
@keegancsmith: Sounds great!

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,6 +25,7 @@ All notable changes to Sourcegraph are documented in this file.
 
 - Search indexer tuned to wait longer before assuming a deadlock has occurred. Previously if the indexserver had many cores (40+) and indexed a monorepo it could give up. [#16110](https://github.com/sourcegraph/sourcegraph/pull/16110)
 - The total size of all Git repositories and the lines of code for indexed branches will be sent back in pings as part of critical telemetry. [#16188](https://github.com/sourcegraph/sourcegraph/pull/16188)
+- The `gitserver` container now has a dependency on Postgres. This does not require any additional configuration unless access to Postgres requires a sidecar proxy / firewall rules. [#16121](https://github.com/sourcegraph/sourcegraph/pull/16121)
```
Comment (keegancsmith) [inline CHANGELOG.md:28]: cc @sourcegraph/distribution gitserver now depends on postgres. I don't believe this requires any updates to documentation/deployments based on how we talk about repo-updater (which has a similiar dependency on postgres). However, we do need to update for cloud due to proxy, see https://github.com/sourcegraph/deploy-sourcegraph-dot-com/pull/3929

**Comment by keegancsmith** [comment]
> @tsenart I can land and shepherd this into production on Monday morning. Then our next hack session we can work on removing cruft in the client side of this :)

Didn't notice the failing tests/etc. Fixed now and did the corresponding PR for k8s. I timeboxed one hour to this and I'm at 1h06m so stopping. Might find time later today to deploy out, but otherwise will deploy later this week.

**Comment by keegancsmith** [comment]
> Didn't notice the failing tests/etc. Fixed now and did the corresponding PR for k8s. I timeboxed one hour to this and I'm at 1h06m so stopping. Might find time later today to deploy out, but otherwise will deploy later this week.

CI was still failing due to a linter check. I got codesniped at improving the go-dbconn-import linter check. Now at 1h18m and stopping :)

**Comment by tsenart** [comment]
@keegancsmith: Shall we force a deploy instead of waiting for Renovate?

**Pull request merged**